src/strings.js

```diff
@@ -40,54 +40,115 @@ function rewritePatternForDotAll(pattern) {
   return cachedRegExp[pattern];
 }
 
-engine.indexOf = function(coll, substr){
+engine.indexOf = function (coll, substr) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(substr) || util.isEmpty(str) ? [] : str.indexOf(substr);
 };
 
-engine.substring = function(coll, start, length){
+engine.substring = function (coll, start, length) {
   const str = misc.singleton(coll, 'String');
   if (util.isEmpty(str) || util.isEmpty(start) || start < 0 || start >= str.length) {
-    return  [];
+    return [];
   }
   if (length === undefined || util.isEmpty(length)) {
     return str.substring(start);
   }
   return str.substring(start, start + length);
 };
 
-engine.startsWith = function(coll, prefix){
+engine.startsWith = function (coll, prefix) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(prefix) || util.isEmpty(str) ? [] : str.startsWith(prefix);
 };
 
-engine.endsWith = function(coll, postfix) {
+engine.endsWith = function (coll, postfix) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(postfix) || util.isEmpty(str) ? [] : str.endsWith(postfix);
 };
 
-engine.containsFn = function(coll, substr){
+engine.containsFn = function (coll, substr) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(substr) || util.isEmpty(str) ? [] : str.includes(substr);
 };
 
-engine.upper = function(coll){
+engine.upper = function (coll) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(str) ? [] : str.toUpperCase();
 };
 
-
-engine.lower = function(coll){
+engine.lower = function (coll) {
   const str = misc.singleton(coll, 'String');
   return util.isEmpty(str) ? [] : str.toLowerCase();
 };
 
+engine.joinFn = function (coll, separator) {
+  const stringValues = coll.map((n) => {
+    const d = util.valData(n);
+    if (typeof d === "string") {
+      return d;
+    }
+    return null;
+  });
+  return stringValues.join(separator);
+};
+
+engine.splitFn = function (coll, separator) {
+  const strToSplit = misc.singleton(coll, 'String');
+  return util.isEmpty(strToSplit) ? [] : strToSplit.split(separator);
+};
+
+engine.trimFn = function (coll) {
+  const stringValues = coll.map((n) => {
+    const d = util.valData(n);
+    if (typeof d === "string") {
+      return d.trim();
+    }
+    return null;
+  });
+  return stringValues;
+};
+
+// encoding/decoding
+engine.encodeFn = function (coll, format) {
+  const strToEncode = misc.singleton(coll, 'String');
+  if (util.isEmpty(strToEncode)){
+    return [];
+  }
+  if (format === 'urlbase64'){
+    return encodeURIComponent(strToEncode);
```

Comment:
It seems that JavaScript's encodeURIComponent leaves some characters unencoded which FHIRPath wants encoded.  From the [FHIRPath PR](https://github.com/HL7/FHIRPath/pull/10/files#diff-4de6dbd4bce188297247b78b84639ec28500b6e90a783241539e95c9e33fe4d0R1448) the resulting output should consist of A-Z, a-z, 0-9, -, and _, but [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) leaves these characters unencoded: A-Z a-z 0-9 - _ . ! ~ * ' ( )